### PR TITLE
tls: Add descriptions for x509v3 alt name verification (tls.verify_hostname)

### PR DESCRIPTION
**Pull request by cosmo0920**
Corresponding to https://github.com/fluent/fluent-bit/pull/8966.

### CODE REVIEW

File: administration/transport-security.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ Both input and output plugins that perform Network I/O can optionally enable TLS
 | :--- | :--- | :--- |
 | tls | enable or disable TLS support | Off |
 | tls.verify | force certificate validation | On |
+| tls.verify\_hostname | force hostname validation for certificates | Off |
```
Comment (patrick-stephens) [inline administration/transport-security.md:12]: Slightly different wording to below.

File: administration/transport-security.md [hunk 2]
```diff
@@ -171,3 +172,41 @@ Fluent Bit supports [TLS server name indication](https://en.wikipedia.org/wiki/S
     tls.ca_file /etc/certs/fluent.crt
     tls.vhost   fluent.example.com
 ```
+
+### Verify subjectAltName
+
+When extracting the X509v3 Subject Alternative Name from a certificate:
```
Comment (patrick-stephens) [inline administration/transport-security.md:178]: I maybe would reword this a bit, using `Then` as well is not ideal.

Something like?
```suggestion.
By default, TLS verification of hostnames is not done automatically.
As an example, we can extract the X509v3 Subject Alternative Name from a certificate:
```

File: administration/transport-security.md [hunk 3]
```diff
@@ -171,3 +172,41 @@ Fluent Bit supports [TLS server name indication](https://en.wikipedia.org/wiki/S
...
+
+```
+X509v3 Subject Alternative Name:
+    DNS:my.fluent-aggregator.net
+```
+
+Then, the TLS client will connect to `other.fluent-aggregator.net`, there is a glitch of hostnames.
```
Comment (patrick-stephens) [inline administration/transport-security.md:185]:
```suggestion
As you can see, this certificate covers only `my.fluent-aggregator.net` so if we use a different hostname it should fail.
```

File: administration/transport-security.md [hunk 4]
```diff
@@ -171,3 +172,41 @@ Fluent Bit supports [TLS server name indication](https://en.wikipedia.org/wiki/S
...
+
+This difference of the hostname can be detected with:
```
Comment (patrick-stephens) [inline administration/transport-security.md:187]:
```suggestion
To fully verify the alternative name and demonstrate the failure we enable `tls.verify_hostname`:
```

**Review by patrick-stephens — COMMENTED** [review]
Is there a reason we default to not verifying the alternative name? This seems counter to an expectation of secure by default.

**Comment by cosmo0920** [comment]
> Is there a reason we default to not verifying the alternative name? This seems counter to an expectation of secure by default.

Once, I implemented this feature as forcibly style, I just though such way. Because hostname verification is mostly required for TLS connections to prevent malformed hostname attached instances. However, we got realized that there is many issues in certificates for kubelet. So, preserving the backward compatibilities, we need to disable by default.

**Comment by patrick-stephens** [comment]
> > Is there a reason we default to not verifying the alternative name? This seems counter to an expectation of secure by default.
> 
> Once, I implemented this feature as a forcibly style, I just though such way. Because hostname verification is mostly required for TLS connections to prevent malformed hostname attached instances. However, we got realized that there is many issues in certificates for kubelet. So, preserving the backward compatibilities, we need to disable by default.

Is this just to handle kubelet specific things then? I'd expect tls verify for everything else, e.g. outputs, etc., so can we just add a special config for kubelet usage to disable it?

**Comment by cosmo0920** [comment]
> > > Is there a reason we default to not verifying the alternative name? This seems counter to an expectation of secure by default.
> > 
> > 
> > Once, I implemented this feature as a forcibly style, I just though such way. Because hostname verification is mostly required for TLS connections to prevent malformed hostname attached instances. However, we got realized that there is many issues in certificates for kubelet. So, preserving the backward compatibilities, we need to disable by default.
> 
> Is this just to handle kubelet specific things then? I'd expect tls verify for everything else, e.g. outputs, etc., so can we just add a special config for kubelet usage to disable it?

Hmm.., I see. For 3.x line, we need to restore the behavior. But it's reasonable to verify certificates as much as possible in 3.1. 
@edsiper Can we break the behavior around certificates in Fluent Bit 3.1? It's reasonable to verify certificates except for filter_kubernetes.

**Comment by cosmo0920** [comment]
Oh, merged my PRs as-is. Also, users complains in in_forward and filter_kubernetes. This is because our network stack is not dumped for its error status with human readable errors. So, we need to restore the behavior. For TLS client, we can print the error message from OpenSSL's certificate verification. But, for server, we didn't get how to print an error with human readable modifications.

**Review by lecaros — APPROVED** [review]
